src/main/scala/akka/persistence/cassandra/journal/StreamMergerActor.scala

```diff
@@ -0,0 +1,134 @@
+package akka.persistence.cassandra.journal
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+import scala.util.control.Exception._
+
+import akka.actor.{Props, Actor}
+import akka.persistence.cassandra.journal.StreamMerger._
+import akka.serialization.SerializationExtension
+import com.datastax.driver.core.Session
+import com.datastax.driver.core.exceptions.DriverException
+
+object StreamMergerActor {
+  def props(config: CassandraJournalConfig, session: Session): Props =
+    Props(new StreamMergerActor(config, session))
+
+  val name = "StreamMerger"
+}
+
+/**
+ * Merges n independent physical streams into a single logical stream.
+ * Preserves causal order per persistenceId.
+ *
+ * The intent of current functionality is to work in batches. The merger retrieves data from all
+ * streams up to a maximum of configured step. Then creates a single logical stream out of the
+ * n independent streams. There is no ordering between the independent streams and the only
+ * requirement is to maintain causal order per persistenceId. When the whole batch can be merged
+ * the stream is emitted. If not (e.g. causality violation in given batch) then the merger
+ * requests again for more data.
+ *
+ * The emmited logical stream can be stored into database in a batch. This grants atomicity, but not
+ * isolation which should be what we need. We can recover with RYOW consistency from the persisted
+ * index table similarly to original functionality. But if we store the progress in all independent
+ * streams we should be able to rerun the merging and idempotency shoud ensure we don't emit
+ * duplicates. If part of the batch is not perceived as complete due to eventual
+ * consistency should be fine and reconciled during replay.
```

Comment:
Sorry I think I mixed a few things together during rewrites. There are a few things that need to be decided. This PR actually only introduces 1), but the others are very important for the next PR (writing to index tables) and also for replay and recovery scenarios.

1) Stream merging component algorithm granularity
a) Update immediatelly when an expected (not out of sequence) element is encountered.
b) Merge the whole batch before updating.
I think this does not matter much. The second approach requires us to keep a bit more in memory, but nicely separates the merging logic from updating the index tables and provides more flexibility which is why it was chosen.

2) Index table updates condition
a) Only update index tables when the batch was merged successfully. 
b) Update even if some elements were out of sequence.

3) Index table update granularity
a) Update each element separately or batch per partition.
b) Batch updates.

Different combinations result in different representations in index table, recovery and replay.

a) (2a, 3b). Store JournalSequenceId of each independent physical stream with each index table write. If we updates in batches and only updated when the streams were merged successfuly we would not have to keep any progress representation except the journalSequenceIds. This scenario seems to be quite similar to the original akka persistence cassandra journal writing functionality. In case of replay we would simply read the events in index table (events by persistence id) for given persistenceId and completed the read by reading from the unmerged journals. We would therefore need to keep track of progress of each independent stream. For recovery we would need to find last write in all partitions in the index table (events by persistence id) and recover from the latest progress. Since the paritions are updated in batches the writes we read may not be the latest, but that should be fine, because the stream merging would only merge some elements twice and store them with the same persistenceId-parirtitionNr-sequenceNr combination which should be idempotent and preserve causality. Does not seem very efficient though.

b) (2a, 3b). Same as above + store latest JournalSequenceId of each independent physical stream separately. Similar to previous, but during recovery stream merging component would not have to scan all index table partitions and it would only look up the latest known progress from a separate table. That table would be updated in a batch with each write. As you mentioned that gives us atomicity, but not isolation, and eventual consistency. That should be fine as merging twice should be idempotent.

c) (2a, 3a). We could update each element separately after the batch is successfully merged. That would be quite nice and would give us RYOW consistency, because we would always only update one partition. But not sure about performance and assuming the stream merging component may fail at any time we need to keep progress similarly to a) or b) so I do not see any real advantage of doing this. A batch per partition could be nice and performance, but since the batch could potentially include elements from multiple physical streams it would be complicated to keep track of progress in case of failure and would likely require cross partition updates anyway.

d) (2b, 3b). Updating even if some elements of the currently merged part of the independent streams were out of order. That means they are in a "buffer" waiting to be merged into the single stream, but we need to load more data to be able to do it. Since however stream merging component may fail at any time keeping it in memory only really is not an option and could result in not emmitted elements. In addition to previously stored information we would need to keep this buffer in database. That can represented in a separate buffer table or for example as a column (boolean) in the original independent streams table. We would have to take that into consideration during recovery/replay which should be doable and actually quite similar to the previous options. Again in worst case during failure because we would be updating multiple partitions in a batch the element could be merged twice which should be ok.

e) (2b, 3a). Similarly to c) with relation to a) and b) I do not see any real advantage of updating elements separately. Unless there is a performance benefit?

With that said I think the most reasonable option is 1b, 2a, 3a/3b, but 3b seems more likely. What do you think?